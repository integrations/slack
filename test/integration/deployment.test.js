const request = require('supertest');
const nock = require('nock');

const { probot, models } = require('.');
const fixtures = require('../fixtures');

const {
  SlackUser,
  GitHubUser,
  Installation,
} = models;

describe('Integration: Creating and listing deployments from Slack', () => {
  let workspace;
  let githubUser;
  beforeEach(async () => {
    const { SlackWorkspace } = models;
    workspace = await SlackWorkspace.create({
      slackId: 'T0001',
      accessToken: 'xoxa-token',
    });

    githubUser = await GitHubUser.create({
      id: 1,
      accessToken: 'secret',
    });

    await SlackUser.create({
      slackId: 'U2147483697', // same as in fixtures.slack.command
      slackWorkspaceId: workspace.id,
      githubId: githubUser.id,
    });

    await Installation.create({
      githubId: 1,
      ownerId: 1337,
    });
  });
  test('works when specifying a repository', async () => {
    nock('https://api.github.com').get('/repos/kubernetes/kubernetes').reply(200, {
      full_name: 'kubernetes/kubernetes',
      id: 54321,
    });
    nock('https://api.github.com').get('/repos/kubernetes/kubernetes/branches').reply(200, fixtures.branches);
    nock('https://api.github.com').get('/repos/kubernetes/kubernetes/git/refs/tags').reply(200, fixtures.tags);

    nock('https://slack.com').post('/api/dialog.open', (body) => {
      expect(body).toMatchSnapshot();
      return true;
    }).reply(200, { ok: true });

    const command = fixtures.slack.command({
      text: 'deploy kubernetes/kubernetes',
    });

    await request(probot.server).post('/slack/command').send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });

    nock('https://api.github.com').get('/repositories/54321').reply(200, {
      name: 'kubernetes',
      owner: {
        login: 'kubernetes',
      },
    });

    nock('https://api.github.com').post('/repos/kubernetes/kubernetes/deployments').reply(200, {
      ref: 'refs/tags/v1.0',
      task: 'deploy',
      environment: 'production',
    });

    // User submits dialog to create a deployment
    await request(probot.server).post('/slack/actions').send({
      payload: JSON.stringify(fixtures.slack.action.dialogSubmissionCreateDeployment()),
    })
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });

  test('works when listing deployments', async () => {
    nock('https://api.github.com').post('/graphql').reply(200, fixtures.deployments);

    const command = fixtures.slack.command({
      text: 'deploy kubernetes/kubernetes list',
    });

    await request(probot.server).post('/slack/command').send(command)
      .expect(200)
      .expect((res) => {
        expect(res.body).toMatchSnapshot();
      });
  });
});
