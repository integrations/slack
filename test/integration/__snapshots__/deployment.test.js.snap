// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: Creating and listing deployments from Slack informs of no deployments found when listing deployments 1`] = `
Object {
  "text": "No deployments found for \`kubernetes/kubernetes\`",
}
`;

exports[`Integration: Creating and listing deployments from Slack informs of no deployments found when listing deployments 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack informs of no deployments found when listing deployments 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes list",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: Creating and listing deployments from Slack informs to accept new permissions 1`] = `
Object {
  "dialog": "{\\"state\\":\\"{\\\\\\"resource\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"repo\\\\\\",\\\\\\"owner\\\\\\":\\\\\\"kubernetes\\\\\\",\\\\\\"repo\\\\\\":\\\\\\"kubernetes\\\\\\"},\\\\\\"channel\\\\\\":\\\\\\"C2147483705\\\\\\"}\\",\\"callback_id\\":\\"create-deployment-dialog\\",\\"title\\":\\"Trigger a deployment\\",\\"submit_label\\":\\"Create\\",\\"elements\\":[{\\"label\\":\\"Repository\\",\\"type\\":\\"select\\",\\"name\\":\\"repository\\",\\"value\\":54321,\\"options\\":[{\\"label\\":\\"kubernetes/kubernetes\\",\\"value\\":54321}]},{\\"label\\":\\"Branch or tag to deploy\\",\\"type\\":\\"select\\",\\"name\\":\\"ref\\",\\"option_groups\\":[{\\"label\\":\\"Tags\\",\\"options\\":[{\\"label\\":\\"refs/tags/v1.12.0-rc.2\\",\\"value\\":\\"refs/tags/v1.12.0-rc.2\\"},{\\"label\\":\\"refs/tags/v1.13.0-alpha.0\\",\\"value\\":\\"refs/tags/v1.13.0-alpha.0\\"}]},{\\"label\\":\\"Branches\\",\\"options\\":[{\\"label\\":\\"release-1.7\\",\\"value\\":\\"release-1.7\\"},{\\"label\\":\\"release-1.8\\",\\"value\\":\\"release-1.8\\"}]}]},{\\"type\\":\\"text\\",\\"label\\":\\"Environment\\",\\"name\\":\\"environment\\",\\"optional\\":true,\\"placeholder\\":\\"Default: \`production\`\\"},{\\"type\\":\\"text\\",\\"label\\":\\"Task\\",\\"name\\":\\"task\\",\\"optional\\":true,\\"placeholder\\":\\"Default: \`deploy\`\\"},{\\"type\\":\\"textarea\\",\\"label\\":\\"Payload\\",\\"name\\":\\"payload\\",\\"placeholder\\":\\"{}\\",\\"hint\\":\\"JSON payload with extra information about the deployment\\",\\"optional\\":true}]}",
  "token": "xoxa-token",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack informs to accept new permissions 2`] = `""`;

exports[`Integration: Creating and listing deployments from Slack informs to accept new permissions 3`] = `
Object {
  "attachments": "[{\\"color\\":\\"danger\\",\\"mrkdwn_in\\":[\\"text\\"],\\"text\\":\\"This feature requires updated permissions.\\\\n\\\\n_Note: If you are not the owner of the repository or the owner of the organization, you will have to ask the owner to install it for you. Give them this link: https://github.com/organizations/integrations/settings/installations/90793/permissions/update _\\",\\"actions\\":[{\\"style\\":\\"primary\\",\\"text\\":\\"Update GitHub App permissions\\",\\"type\\":\\"button\\",\\"url\\":\\"https://github.com/organizations/integrations/settings/installations/90793/permissions/update\\"}]}]",
  "channel": "C2147483705",
  "response_type": "ephemeral",
  "token": "xoxa-token",
}
`;

exports[`Integration: Creating and listing deployments from Slack informs to accept new permissions 4`] = `Object {}`;

exports[`Integration: Creating and listing deployments from Slack informs to accept new permissions 5`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack informs to accept new permissions 6`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: Creating and listing deployments from Slack throws error when listing deployments and none are found 1`] = `
Object {
  "attachments": Array [
    Object {
      "actions": Array [
        Object {
          "style": "primary",
          "text": "Update GitHub App permissions",
          "type": "button",
          "url": "https://github.com/organizations/integrations/settings/installations/90793/permissions/update",
        },
      ],
      "color": "danger",
      "mrkdwn_in": Array [
        "text",
      ],
      "text": "This feature requires updated permissions.

_Note: If you are not the owner of the repository or the owner of the organization, you will have to ask the owner to install it for you. Give them this link: https://github.com/organizations/integrations/settings/installations/90793/permissions/update _",
    },
  ],
  "response_type": "ephemeral",
}
`;

exports[`Integration: Creating and listing deployments from Slack throws error when listing deployments and none are found 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack throws error when listing deployments and none are found 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes list",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: Creating and listing deployments from Slack throws error when specifying a repository with no tags and didnt receive a 404 1`] = `
Object {
  "response_type": "in_channel",
}
`;

exports[`Integration: Creating and listing deployments from Slack throws error when specifying a repository with no tags and didnt receive a 404 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack throws error when specifying a repository with no tags and didnt receive a 404 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: Creating and listing deployments from Slack works when listing deployments 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_icon": "https://avatars.githubusercontent.com/gimenete",
      "author_link": "https://github.com/gimenete",
      "author_name": "gimenete",
      "color": "#24292f",
      "fallback": "gimenete triggered a deploy on production from foobar",
      "fields": Array [
        Object {
          "short": true,
          "title": "Task",
          "value": "deploy",
        },
        Object {
          "short": true,
          "title": "Environment",
          "value": "production",
        },
      ],
      "footer": "kubernetes/kubernetes",
      "footer_icon": "https://assets-cdn.github.com/favicon.ico",
      "pretext": ":rocket: Deployment created by gimenete",
      "title": "foobar - adsf",
      "title_link": "https://github.com/Test-slack-github/test2/commit/c877632a9abfe07c9d0866dc0de2d69a975b5469",
      "ts": 1537454125,
    },
  ],
}
`;

exports[`Integration: Creating and listing deployments from Slack works when listing deployments 2`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when listing deployments 3`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes list",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository 1`] = `
Object {
  "dialog": "{\\"state\\":\\"{\\\\\\"resource\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"repo\\\\\\",\\\\\\"owner\\\\\\":\\\\\\"kubernetes\\\\\\",\\\\\\"repo\\\\\\":\\\\\\"kubernetes\\\\\\"},\\\\\\"channel\\\\\\":\\\\\\"C2147483705\\\\\\"}\\",\\"callback_id\\":\\"create-deployment-dialog\\",\\"title\\":\\"Trigger a deployment\\",\\"submit_label\\":\\"Create\\",\\"elements\\":[{\\"label\\":\\"Repository\\",\\"type\\":\\"select\\",\\"name\\":\\"repository\\",\\"value\\":54321,\\"options\\":[{\\"label\\":\\"kubernetes/kubernetes\\",\\"value\\":54321}]},{\\"label\\":\\"Branch or tag to deploy\\",\\"type\\":\\"select\\",\\"name\\":\\"ref\\",\\"option_groups\\":[{\\"label\\":\\"Tags\\",\\"options\\":[{\\"label\\":\\"refs/tags/v1.12.0-rc.2\\",\\"value\\":\\"refs/tags/v1.12.0-rc.2\\"},{\\"label\\":\\"refs/tags/v1.13.0-alpha.0\\",\\"value\\":\\"refs/tags/v1.13.0-alpha.0\\"}]},{\\"label\\":\\"Branches\\",\\"options\\":[{\\"label\\":\\"release-1.7\\",\\"value\\":\\"release-1.7\\"},{\\"label\\":\\"release-1.8\\",\\"value\\":\\"release-1.8\\"}]}]},{\\"type\\":\\"text\\",\\"label\\":\\"Environment\\",\\"name\\":\\"environment\\",\\"optional\\":true,\\"placeholder\\":\\"Default: \`production\`\\"},{\\"type\\":\\"text\\",\\"label\\":\\"Task\\",\\"name\\":\\"task\\",\\"optional\\":true,\\"placeholder\\":\\"Default: \`deploy\`\\"},{\\"type\\":\\"textarea\\",\\"label\\":\\"Payload\\",\\"name\\":\\"payload\\",\\"placeholder\\":\\"{}\\",\\"hint\\":\\"JSON payload with extra information about the deployment\\",\\"optional\\":true}]}",
  "token": "xoxa-token",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository 2`] = `""`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository 3`] = `
Object {
  "attachments": "[{\\"fallback\\":\\"gimenete triggered a deploy on production from foobar\\",\\"color\\":\\"#24292f\\",\\"pretext\\":\\":rocket: Deployment created by gimenete\\",\\"author_name\\":\\"gimenete\\",\\"author_link\\":\\"https://github.com/gimenete\\",\\"author_icon\\":\\"https://avatars.githubusercontent.com/gimenete\\",\\"title\\":\\"foobar - adsf\\",\\"title_link\\":\\"https://github.com/Test-slack-github/test2/commit/c877632a9abfe07c9d0866dc0de2d69a975b5469\\",\\"fields\\":[{\\"title\\":\\"Task\\",\\"value\\":\\"deploy\\",\\"short\\":true},{\\"title\\":\\"Environment\\",\\"value\\":\\"production\\",\\"short\\":true}],\\"footer\\":\\"kubernetes/kubernetes\\",\\"footer_icon\\":\\"https://assets-cdn.github.com/favicon.ico\\",\\"ts\\":1537454125}]",
  "channel": "C2147483705",
  "token": "xoxa-token",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository 4`] = `Object {}`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository 5`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository 6`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository with no tags 1`] = `
Object {
  "dialog": "{\\"state\\":\\"{\\\\\\"resource\\\\\\":{\\\\\\"type\\\\\\":\\\\\\"repo\\\\\\",\\\\\\"owner\\\\\\":\\\\\\"kubernetes\\\\\\",\\\\\\"repo\\\\\\":\\\\\\"kubernetes\\\\\\"},\\\\\\"channel\\\\\\":\\\\\\"C2147483705\\\\\\"}\\",\\"callback_id\\":\\"create-deployment-dialog\\",\\"title\\":\\"Trigger a deployment\\",\\"submit_label\\":\\"Create\\",\\"elements\\":[{\\"label\\":\\"Repository\\",\\"type\\":\\"select\\",\\"name\\":\\"repository\\",\\"value\\":54321,\\"options\\":[{\\"label\\":\\"kubernetes/kubernetes\\",\\"value\\":54321}]},{\\"label\\":\\"Branch or tag to deploy\\",\\"type\\":\\"select\\",\\"name\\":\\"ref\\",\\"option_groups\\":[{\\"label\\":\\"Branches\\",\\"options\\":[{\\"label\\":\\"release-1.7\\",\\"value\\":\\"release-1.7\\"},{\\"label\\":\\"release-1.8\\",\\"value\\":\\"release-1.8\\"}]}]},{\\"type\\":\\"text\\",\\"label\\":\\"Environment\\",\\"name\\":\\"environment\\",\\"optional\\":true,\\"placeholder\\":\\"Default: \`production\`\\"},{\\"type\\":\\"text\\",\\"label\\":\\"Task\\",\\"name\\":\\"task\\",\\"optional\\":true,\\"placeholder\\":\\"Default: \`deploy\`\\"},{\\"type\\":\\"textarea\\",\\"label\\":\\"Payload\\",\\"name\\":\\"payload\\",\\"placeholder\\":\\"{}\\",\\"hint\\":\\"JSON payload with extra information about the deployment\\",\\"optional\\":true}]}",
  "token": "xoxa-token",
  "trigger_id": "13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository with no tags 2`] = `""`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository with no tags 3`] = `
Object {
  "attachments": "[{\\"fallback\\":\\"gimenete triggered a deploy on production from foobar\\",\\"color\\":\\"#24292f\\",\\"pretext\\":\\":rocket: Deployment created by gimenete\\",\\"author_name\\":\\"gimenete\\",\\"author_link\\":\\"https://github.com/gimenete\\",\\"author_icon\\":\\"https://avatars.githubusercontent.com/gimenete\\",\\"title\\":\\"foobar - adsf\\",\\"title_link\\":\\"https://github.com/Test-slack-github/test2/commit/c877632a9abfe07c9d0866dc0de2d69a975b5469\\",\\"fields\\":[{\\"title\\":\\"Task\\",\\"value\\":\\"deploy\\",\\"short\\":true},{\\"title\\":\\"Environment\\",\\"value\\":\\"production\\",\\"short\\":true}],\\"footer\\":\\"kubernetes/kubernetes\\",\\"footer_icon\\":\\"https://assets-cdn.github.com/favicon.ico\\",\\"ts\\":1537454125}]",
  "channel": "C2147483705",
  "token": "xoxa-token",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository with no tags 4`] = `Object {}`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository with no tags 5`] = `
Set {
  "pending-command#13345224609.738474920.8088930838d88f008e0",
}
`;

exports[`Integration: Creating and listing deployments from Slack works when specifying a repository with no tags 6`] = `
Map {
  "pending-command#13345224609.738474920.8088930838d88f008e0" => Object {
    "channel_id": "C2147483705",
    "channel_name": "test",
    "command": "/github",
    "enterprise_id": "E0001",
    "enterprise_name": "Globular%20Construct%20Inc",
    "response_url": "https://hooks.slack.com/commands/1234/5678",
    "team_domain": "example",
    "team_id": "T0001",
    "text": "deploy kubernetes/kubernetes",
    "token": "secret",
    "trigger_id": "13345224609.738474920.8088930838d88f008e0",
    "user_id": "U2147483697",
    "user_name": "Steve",
  },
}
`;
